File: frontend/src/lib/voice-engine/realtime-client.ts
/**
 * OpenAI Realtime API WebSocket client for voice conversations.
 *
 * Connects to wss://api.openai.com/v1/realtime and manages the full
 * lifecycle of a voice session: audio streaming, VAD, tool calls,
 * and response playback.
 */

export type RealtimeVoice =
  | "alloy"
  | "echo"
  | "fable"
  | "onyx"
  | "nova"
  | "shimmer";

export interface VADConfig {
  /** "server_vad" lets OpenAI detect speech boundaries. */
  type: "server_vad";
  /** Silence duration (ms) before end-of-turn. Default 500. */
  silence_duration_ms?: number;
  /** Audio level threshold (0-1). Default 0.5. */
  threshold?: number;
  /** Minimum speech duration (ms) to count as speech. Default 200. */
  prefix_padding_ms?: number;
}

export interface RealtimeToolDefinition {
  type: "function";
  name: string;
  description: string;
  parameters: Record<string, unknown>;
}

export interface RealtimeSessionConfig {
  model?: string;
  voice?: RealtimeVoice;
  instructions?: string;
  tools?: RealtimeToolDefinition[];
  input_audio_format?: "pcm16" | "g711_ulaw" | "g711_alaw";
  output_audio_format?: "pcm16" | "g711_ulaw" | "g711_alaw";
  input_audio_transcription?: { model: string };
  turn_detection?: VADConfig | null;
  temperature?: number;
  max_response_output_tokens?: number | "inf";
}

export type RealtimeEvent =
  | { type: "session.created"; session: Record<string, unknown> }
  | { type: "session.updated"; session: Record<string, unknown> }
  | { type: "input_audio_buffer.speech_started" }
  | { type: "input_audio_buffer.speech_stopped" }
  | { type: "input_audio_buffer.committed" }
  | { type: "conversation.item.created"; item: Record<string, unknown> }
  | {
      type: "response.audio.delta";
      delta: string;
      response_id: string;
      item_id: string;
    }
  | { type: "response.audio.done"; response_id: string; item_id: string }
  | {
      type: "response.audio_transcript.delta";
      delta: string;
      response_id: string;
      item_id: string;
    }
  | {
      type: "response.audio_transcript.done";
      transcript: string;
      response_id: string;
      item_id: string;
    }
  | {
      type: "response.function_call_arguments.delta";
      delta: string;
      call_id: string;
      name: string;
    }
  | {
      type: "response.function_call_arguments.done";
      arguments: string;
      call_id: string;
      name: string;
    }
  | { type: "response.done"; response: Record<string, unknown> }
  | {
      type: "input_audio_buffer.transcription.completed";
      transcript: string;
    }
  | { type: "error"; error: { message: string; type: string; code?: string } }
  | { type: string; [key: string]: unknown };

export interface RealtimeClientCallbacks {
  onOpen?: () => void;
  onClose?: (code: number, reason: string) => void;
  onError?: (error: Error) => void;
  onEvent?: (event: RealtimeEvent) => void;
  onAudioDelta?: (base64Audio: string) => void;
  onTranscriptDelta?: (delta: string, role: "assistant") => void;
  onTranscriptDone?: (transcript: string, role: "assistant") => void;
  onInputTranscript?: (transcript: string) => void;
  onSpeechStarted?: () => void;
  onSpeechStopped?: () => void;
  onFunctionCall?: (
    name: string,
    args: string,
    callId: string
  ) => Promise<string>;
  onResponseDone?: (response: Record<string, unknown>) => void;
}

const REALTIME_URL = "wss://api.openai.com/v1/realtime";
const DEFAULT_MODEL = "gpt-4o-realtime-preview";

export class RealtimeClient {
  private ws: WebSocket | null = null;
  private apiKey: string;
  private config: RealtimeSessionConfig;
  private callbacks: RealtimeClientCallbacks;
  private _connected = false;
  private _sessionId: string | null = null;
  private autoEndTimer: ReturnType<typeof setTimeout> | null = null;
  private maxDurationMs: number;

  constructor(opts: {
    apiKey: string;
    config: RealtimeSessionConfig;
    callbacks: RealtimeClientCallbacks;
    /** Max call duration in minutes. Default 20. */
    maxDurationMinutes?: number;
  }) {
    this.apiKey = opts.apiKey;
    this.config = {
      model: DEFAULT_MODEL,
      voice: "alloy",
      input_audio_format: "pcm16",
      output_audio_format: "pcm16",
      input_audio_transcription: { model: "whisper-1" },
      turn_detection: {
        type: "server_vad",
        silence_duration_ms: 500,
        threshold: 0.5,
        prefix_padding_ms: 300,
      },
      temperature: 0.7,
      max_response_output_tokens: 4096,
      ...opts.config,
    };
    this.callbacks = opts.callbacks;
    this.maxDurationMs = (opts.maxDurationMinutes ?? 20) * 60 * 1000;
  }

  get connected(): boolean {
    return this._connected;
  }
  get sessionId(): string | null {
    return this._sessionId;
  }

  /** Open the WebSocket and configure the session. */
  connect(): void {
    const model = this.config.model ?? DEFAULT_MODEL;
    const url = `${REALTIME_URL}?model=${encodeURIComponent(model)}`;

    this.ws = new WebSocket(url, [
      "realtime",
      `openai-insecure-api-key.${this.apiKey}`,
      "openai-beta.realtime-v1",
    ]);

    this.ws.onopen = () => {
      this._connected = true;
      this.callbacks.onOpen?.();
      this.updateSession();
      // Auto-end after max duration
      this.autoEndTimer = setTimeout(() => this.disconnect(), this.maxDurationMs);
    };

    this.ws.onclose = (e) => {
      this._connected = false;
      this.clearAutoEnd();
      this.callbacks.onClose?.(e.code, e.reason);
    };

    this.ws.onerror = () => {
      this.callbacks.onError?.(new Error("WebSocket error"));
    };

    this.ws.onmessage = (e) => {
      try {
        const event: RealtimeEvent = JSON.parse(
          typeof e.data === "string" ? e.data : ""
        );
        this.handleEvent(event);
      } catch {
        // ignore unparseable frames
      }
    };
  }

  /** Gracefully close the connection. */
  disconnect(): void {
    this.clearAutoEnd();
    if (this.ws) {
      this.ws.close(1000, "Client disconnect");
      this.ws = null;
    }
    this._connected = false;
  }

  /** Send raw event to the server. */
  send(event: Record<string, unknown>): void {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) return;
    this.ws.send(JSON.stringify(event));
  }

  /** Push a chunk of PCM16 audio (base64-encoded) into the input buffer. */
  appendAudio(base64Audio: string): void {
    this.send({
      type: "input_audio_buffer.append",
      audio: base64Audio,
    });
  }

  /** Manually commit the current audio buffer (useful if VAD is off). */
  commitAudio(): void {
    this.send({ type: "input_audio_buffer.commit" });
  }

  /** Clear pending input audio. */
  clearAudio(): void {
    this.send({ type: "input_audio_buffer.clear" });
  }

  /** Interrupt in-progress response (barge-in). */
  cancelResponse(): void {
    this.send({ type: "response.cancel" });
  }

  /** Send a text message as user input. */
  sendText(text: string): void {
    this.send({
      type: "conversation.item.create",
      item: {
        type: "message",
        role: "user",
        content: [{ type: "input_text", text }],
      },
    });
    this.send({ type: "response.create" });
  }

  /** Submit function call result back to the model. */
  submitFunctionResult(callId: string, result: string): void {
    this.send({
      type: "conversation.item.create",
      item: {
        type: "function_call_output",
        call_id: callId,
        output: result,
      },
    });
    this.send({ type: "response.create" });
  }

  /** Update the session configuration on the fly. */
  updateSession(overrides?: Partial<RealtimeSessionConfig>): void {
    const session: Record<string, unknown> = {};
    const merged = { ...this.config, ...overrides };

    if (merged.instructions) session.instructions = merged.instructions;
    if (merged.voice) session.voice = merged.voice;
    if (merged.tools) session.tools = merged.tools;
    if (merged.input_audio_format)
      session.input_audio_format = merged.input_audio_format;
    if (merged.output_audio_format)
      session.output_audio_format = merged.output_audio_format;
    if (merged.input_audio_transcription)
      session.input_audio_transcription = merged.input_audio_transcription;
    if (merged.turn_detection !== undefined)
      session.turn_detection = merged.turn_detection;
    if (merged.temperature !== undefined)
      session.temperature = merged.temperature;
    if (merged.max_response_output_tokens !== undefined)
      session.max_response_output_tokens = merged.max_response_output_tokens;

    this.send({ type: "session.update", session });
    if (overrides) Object.assign(this.config, overrides);
  }

  // ── Private ────────────────────────────────────────────────────────────────

  private clearAutoEnd(): void {
    if (this.autoEndTimer) {
      clearTimeout(this.autoEndTimer);
      this.autoEndTimer = null;
    }
  }

  private async handleEvent(event: RealtimeEvent): Promise<void> {
    this.callbacks.onEvent?.(event);

    const e = event as Record<string, unknown>;

    switch (event.type) {
      case "session.created": {
        const session = e.session as Record<string, unknown> | undefined;
        this._sessionId = (session?.id as string) ?? null;
        break;
      }

      case "input_audio_buffer.speech_started":
        this.callbacks.onSpeechStarted?.();
        break;

      case "input_audio_buffer.speech_stopped":
        this.callbacks.onSpeechStopped?.();
        break;

      case "response.audio.delta":
        this.callbacks.onAudioDelta?.(e.delta as string);
        break;

      case "response.audio_transcript.delta":
        this.callbacks.onTranscriptDelta?.(e.delta as string, "assistant");
        break;

      case "response.audio_transcript.done":
        this.callbacks.onTranscriptDone?.(e.transcript as string, "assistant");
        break;

      case "input_audio_buffer.transcription.completed":
        this.callbacks.onInputTranscript?.(e.transcript as string);
        break;

      case "response.function_call_arguments.done":
        if (this.callbacks.onFunctionCall) {
          const fnName = e.name as string;
          const fnArgs = e.arguments as string;
          const fnCallId = e.call_id as string;
          try {
            const result = await this.callbacks.onFunctionCall(
              fnName,
              fnArgs,
              fnCallId
            );
            this.submitFunctionResult(fnCallId, result);
          } catch (err) {
            const msg =
              err instanceof Error ? err.message : "Function call failed";
            this.submitFunctionResult(
              fnCallId,
              JSON.stringify({ error: msg })
            );
          }
        }
        break;

      case "response.done":
        this.callbacks.onResponseDone?.(
          e.response as Record<string, unknown>
        );
        break;

      case "error":
        this.callbacks.onError?.(
          new Error(
            ((e.error as Record<string, unknown>)?.message as string) ??
              "Unknown error"
          )
        );
        break;
    }
  }
}

File: backend/app/services/voice/realtime_client.py
"""OpenAI Realtime API WebSocket client for voice conversations."""

from __future__ import annotations

import asyncio
import audioop
import base64
import json
import logging
from collections.abc import AsyncGenerator, Callable
from dataclasses import dataclass, field
from enum import Enum
from typing import Any

import websockets
from tenacity import (
    retry,
    retry_if_exception_type,
    stop_after_attempt,
    wait_exponential,
)
from websockets.exceptions import ConnectionClosed, WebSocketException

from app.core.config import get_settings

logger = logging.getLogger(__name__)

REALTIME_API_URL = "wss://api.openai.com/v1/realtime"
REALTIME_MODEL = "gpt-4o-realtime-preview-2024-10-01"

# Audio constants
TWILIO_SAMPLE_RATE = 8000   # μ-law 8kHz from Twilio
OPENAI_SAMPLE_RATE = 24000  # PCM 16-bit 24kHz for OpenAI
AUDIO_CHANNELS = 1


class RealtimeEventType(str, Enum):
    # Server → client
    SESSION_CREATED = "session.created"
    SESSION_UPDATED = "session.updated"
    RESPONSE_AUDIO_DELTA = "response.audio.delta"
    RESPONSE_AUDIO_TRANSCRIPT_DELTA = "response.audio_transcript.delta"
    RESPONSE_DONE = "response.done"
    INPUT_AUDIO_BUFFER_SPEECH_STARTED = "input_audio_buffer.speech_started"
    INPUT_AUDIO_BUFFER_SPEECH_STOPPED = "input_audio_buffer.speech_stopped"
    RESPONSE_FUNCTION_CALL_ARGUMENTS_DONE = "response.function_call_arguments.done"
    ERROR = "error"
    # Client → server
    SESSION_UPDATE = "session.update"
    INPUT_AUDIO_BUFFER_APPEND = "input_audio_buffer.append"
    INPUT_AUDIO_BUFFER_COMMIT = "input_audio_buffer.commit"
    RESPONSE_CREATE = "response.create"
    CONVERSATION_ITEM_CREATE = "conversation.item.create"


@dataclass
class VADConfig:
    type: str = "server_vad"
    threshold: float = 0.5
    prefix_padding_ms: int = 300
    silence_duration_ms: int = 500


@dataclass
class SessionConfig:
    voice: str = "alloy"
    system_prompt: str = ""
    vad: VADConfig = field(default_factory=VADConfig)
    tools: list[dict[str, Any]] = field(default_factory=list)
    temperature: float = 0.8
    max_response_output_tokens: int | str = "inf"


EventHandler = Callable[[dict[str, Any]], None]


def ulaw_to_pcm24k(ulaw_bytes: bytes) -> bytes:
    """Convert Twilio μ-law 8kHz audio to PCM 16-bit 24kHz for OpenAI."""
    # μ-law → linear PCM 16-bit at 8kHz
    pcm_8k = audioop.ulaw2lin(ulaw_bytes, 2)
    # Resample 8kHz → 24kHz (3x upsample)
    pcm_24k, _ = audioop.ratecv(pcm_8k, 2, AUDIO_CHANNELS, TWILIO_SAMPLE_RATE, OPENAI_SAMPLE_RATE, None)
    return pcm_24k


def pcm24k_to_ulaw(pcm_bytes: bytes) -> bytes:
    """Convert OpenAI PCM 16-bit 24kHz audio to Twilio μ-law 8kHz."""
    # Resample 24kHz → 8kHz
    pcm_8k, _ = audioop.ratecv(pcm_bytes, 2, AUDIO_CHANNELS, OPENAI_SAMPLE_RATE, TWILIO_SAMPLE_RATE, None)
    # Linear PCM → μ-law
    return audioop.lin2ulaw(pcm_8k, 2)


class RealtimeClient:
    """WebSocket client for the OpenAI Realtime API.

    Manages a single bidirectional audio session. Audio flows:
        Twilio (μ-law 8kHz) → ulaw_to_pcm24k → OpenAI (PCM 16-bit 24kHz)
        OpenAI (PCM 16-bit 24kHz) → pcm24k_to_ulaw → Twilio (μ-law 8kHz)
    """

    def __init__(self, session_id: str, config: SessionConfig) -> None:
        self.session_id = session_id
        self.config = config
        self._ws: websockets.WebSocketClientProtocol | None = None
        self._event_handlers: dict[str, list[EventHandler]] = {}
        self._connected = asyncio.Event()
        self._closed = False
        self._send_lock = asyncio.Lock()

    # ------------------------------------------------------------------
    # Connection lifecycle
    # ------------------------------------------------------------------

    @retry(
        retry=retry_if_exception_type((WebSocketException, ConnectionError, OSError)),
        wait=wait_exponential(multiplier=1, min=1, max=30),
        stop=stop_after_attempt(5),
        reraise=True,
    )
    async def connect(self) -> None:
        """Open the WebSocket connection and start the receive loop."""
        settings = get_settings()
        url = f"{REALTIME_API_URL}?model={REALTIME_MODEL}"
        headers = {
            "Authorization": f"Bearer {settings.openai_api_key}",
            "OpenAI-Beta": "realtime=v1",
        }
        logger.info("Connecting to OpenAI Realtime API (session=%s)", self.session_id)
        self._ws = await websockets.connect(url, additional_headers=headers)
        self._closed = False
        self._connected.set()
        asyncio.create_task(self._receive_loop())

    async def close(self) -> None:
        """Close the WebSocket connection gracefully."""
        self._closed = True
        self._connected.clear()
        if self._ws and not self._ws.closed:
            await self._ws.close()
            logger.info("Realtime session closed (session=%s)", self.session_id)

    # ------------------------------------------------------------------
    # Sending events
    # ------------------------------------------------------------------

    async def _send(self, event: dict[str, Any]) -> None:
        if not self._ws or self._ws.closed:
            raise ConnectionError("WebSocket not connected")
        async with self._send_lock:
            await self._ws.send(json.dumps(event))

    async def configure_session(self) -> None:
        """Send session.update with voice, VAD, system prompt, and tools."""
        vad = self.config.vad
        await self._send({
            "type": RealtimeEventType.SESSION_UPDATE,
            "session": {
                "modalities": ["text", "audio"],
                "instructions": self.config.system_prompt,
                "voice": self.config.voice,
                "input_audio_format": "pcm16",
                "output_audio_format": "pcm16",
                "input_audio_transcription": {"model": "whisper-1"},
                "turn_detection": {
                    "type": vad.type,
                    "threshold": vad.threshold,
                    "prefix_padding_ms": vad.prefix_padding_ms,
                    "silence_duration_ms": vad.silence_duration_ms,
                },
                "tools": self.config.tools,
                "tool_choice": "auto",
                "temperature": self.config.temperature,
                "max_response_output_tokens": self.config.max_response_output_tokens,
            },
        })

    async def append_audio(self, ulaw_bytes: bytes) -> None:
        """Accept raw μ-law 8kHz bytes from Twilio and forward to OpenAI."""
        pcm_bytes = ulaw_to_pcm24k(ulaw_bytes)
        encoded = base64.b64encode(pcm_bytes).decode("utf-8")
        await self._send({
            "type": RealtimeEventType.INPUT_AUDIO_BUFFER_APPEND,
            "audio": encoded,
        })

    async def commit_audio(self) -> None:
        """Commit the audio buffer (manual VAD mode)."""
        await self._send({"type": RealtimeEventType.INPUT_AUDIO_BUFFER_COMMIT})

    async def request_response(self) -> None:
        """Trigger a model response (manual turn mode)."""
        await self._send({"type": RealtimeEventType.RESPONSE_CREATE})

    async def send_function_result(self, call_id: str, result: str) -> None:
        """Send a tool/function call result back to the model."""
        await self._send({
            "type": RealtimeEventType.CONVERSATION_ITEM_CREATE,
            "item": {
                "type": "function_call_output",
                "call_id": call_id,
                "output": result,
            },
        })
        await self.request_response()

    # ------------------------------------------------------------------
    # Receiving events
    # ------------------------------------------------------------------

    async def _receive_loop(self) -> None:
        """Continuously read from the WebSocket and dispatch events."""
        try:
            async for raw in self._ws:
                if self._closed:
                    break
                try:
                    event = json.loads(raw)
                    await self._dispatch(event)
                except json.JSONDecodeError:
                    logger.warning("Received non-JSON from Realtime API (session=%s)", self.session_id)
        except ConnectionClosed as exc:
            logger.info("Realtime WebSocket closed (session=%s, code=%s)", self.session_id, exc.code)
        except Exception:
            logger.exception("Unexpected error in Realtime receive loop (session=%s)", self.session_id)
        finally:
            self._connected.clear()

    async def _dispatch(self, event: dict[str, Any]) -> None:
        event_type = event.get("type", "")
        handlers = self._event_handlers.get(event_type, []) + self._event_handlers.get("*", [])
        for handler in handlers:
            try:
                if asyncio.iscoroutinefunction(handler):
                    await handler(event)
                else:
                    handler(event)
            except Exception:
                logger.exception(
                    "Error in event handler for %s (session=%s)", event_type, self.session_id
                )

    # ------------------------------------------------------------------
    # Event handler registration
    # ------------------------------------------------------------------

    def on(self, event_type: str, handler: EventHandler) -> None:
        """Register a handler for a specific event type (or '*' for all)."""
        self._event_handlers.setdefault(event_type, []).append(handler)

    def off(self, event_type: str, handler: EventHandler) -> None:
        handlers = self._event_handlers.get(event_type, [])
        if handler in handlers:
            handlers.remove(handler)

    # ------------------------------------------------------------------
    # Helpers
    # ------------------------------------------------------------------

    async def audio_chunks(self) -> AsyncGenerator[bytes, None]:
        """Yield PCM 24kHz audio chunks decoded from response.audio.delta events."""
        queue: asyncio.Queue[bytes | None] = asyncio.Queue()

        async def _on_audio(event: dict[str, Any]) -> None:
            delta = event.get("delta", "")
            if delta:
                queue.put_nowait(base64.b64decode(delta))

        async def _on_done(_: dict[str, Any]) -> None:
            queue.put_nowait(None)

        self.on(RealtimeEventType.RESPONSE_AUDIO_DELTA, _on_audio)
        self.on(RealtimeEventType.RESPONSE_DONE, _on_done)
        try:
            while True:
                chunk = await queue.get()
                if chunk is None:
                    break
                yield chunk
        finally:
            self.off(RealtimeEventType.RESPONSE_AUDIO_DELTA, _on_audio)
            self.off(RealtimeEventType.RESPONSE_DONE, _on_done)

    @property
    def is_connected(self) -> bool:
        return self._ws is not None and not self._ws.closed and not self._closed
